### PR TITLE
Refresh "profiling with Tracy" developer docs.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/openxla/iree/issues/14470

|  |  |
| -- | -- |
| Current page | https://iree.dev/developers/performance/profiling-with-tracy/ |
| This PR | https://scotttodd.github.io/iree/developers/performance/profiling-with-tracy/ |

This restructures the "Profiling with Tracy" developer doc / website page to focus less on the specifics of build and compile configurations and more on what Tracy is and how to use it. Troubleshooting tips and building from source instructions are still included but they are off to the side. I also added new screenshots, diagrams, tips, and formatting.

Some of the "building from source" instructions can be further simplified if we
* start using [Tracy's recently added upstream CMake build](https://github.com/wolfpld/tracy/blob/master/CMakeLists.txt) (in fact, the old Visual Studio solutions and Makefiles were removed in https://github.com/wolfpld/tracy/commit/7f8a2e2c9a8581c2263da77222f1194ee0653e48)
* address https://github.com/openxla/iree/issues/16777

### CODE REVIEW

**Comment by bjacob** [comment]
Love the restructuring around the higher level overview of Tracy concepts.

Tracy docs need to serve a dual purpose:
1. Explain the concepts - this PR is a nice improvement there.
2. Minimize headaches caused by missing a step. The old state of the doc here reflected a concern for that. Mainly, if a piece of information such as "on desktop Linux, you must run as root" is vital to get good results, then it should be mentioned upfront in the main line of the explanation, and not deferred to an appendix.
    * As this gets heavy, and most of it is only needed for Sampling and not for Instrumentation, and Sampling is enough for many purposes, the old state of the doc tried to find ways to convey that information without obscuring the simpler instructions for Instrumentation-only.
    * The big table at the start of the old doc served those two purposes (make it easy to check that one had not missed a step for sampling, and convey the simpler instructions for Instrumentation vs heavier instructions for Sampling).

File: docs/website/docs/developers/performance/profiling-with-tracy.md [hunk 1]
```diff
@@ -4,72 +4,327 @@ icon: material/chart-line
...
-[view in browser](https://docs.google.com/viewer?url=https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf).
-
-## Overview
-
-We will go through each steps below, but here is an overview. It highlights the
-simpler subset of instructions when only instrumentation is needed, vs. the
-additional steps needed when Sampling is also wanted.
-
-Component | Instrumentation only | Instrumentation and Sampling
---- | --- | ---
```
Comment (ScottTodd) [inline docs/website/docs/developers/performance/profiling-with-tracy.md:43]: > The big table at the start of the old doc served those two purposes (make it easy to check that one had not missed a step for sampling, and convey the simpler instructions for Instrumentation vs heavier instructions for Sampling).

I used to like the large table, but I think it's grown too big. For someone getting started, it's a large information dump that doesn't make it clear if the options are independent from one another.

Specifically, I wanted to:
* Remove the "Build Tracy capture" and "Build Tracy profiler" rows and lean on the upstream build + documentation there. If we're not quite ready for that due to capstone and other Linux build woes, then I'll settle for an appendix section in our docs.
* Simplify "Build the IREE compiler" and "Build IREE device binaries" - All people should really need is a standard source build with `-DIREE_ENABLE_RUNTIME_TRACING` or the Python bindings from pip.

I think the new "capturing a trace" and "quickstart" sections https://scotttodd.github.io/iree/developers/performance/profiling-with-tracy/#capturing-a-trace
cover most of the setup and then people can branch from there.

I'll try reordering sections a bit and will make sure each usage style is explained well enough.

File: docs/website/docs/developers/performance/profiling-with-tracy.md [hunk 2]
```diff
@@ -4,72 +4,327 @@ icon: material/chart-line
...
+2. Compile a program to profile:
+
+    ```shell
+    iree-compile program_input.mlir \
+      --iree-hal-target-backends={target} \
+      --iree-hal-executable-debug-level=3 \
+      -o program.vmfb
+    ```
+
+3. Run the program using the instrumented `iree-run-module`
```
Comment (ScottTodd) [inline docs/website/docs/developers/performance/profiling-with-tracy.md]: >  Mainly, if a piece of information such as "on desktop Linux, you must run as root" is vital to get good results, then it should be mentioned upfront in the main line of the explanation, and not deferred to an appendix.

Noted. I'll iterate on this.

File: docs/website/docs/developers/performance/profiling-with-tracy.md (hunk 1 above)
Comment (ScottTodd) [reply docs/website/docs/developers/performance/profiling-with-tracy.md:43]: @bjacob I made a few changes. Curious to hear your thoughts.

image

Mainly I moved the "Including more information" tips from a section below to inlined example settings with annotations. I also highlighted some key lines in each code block.

Comment (ScottTodd) [reply docs/website/docs/developers/performance/profiling-with-tracy.md:43]: I guess we could also add a more explicit section on instrumentation vs sampling so people don't go down the full sampling path if they aren't particularly looking to use that data.

Comment (bjacob) [reply docs/website/docs/developers/performance/profiling-with-tracy.md:43]: Thanks! I found the little `+` icons (neat -- I didn't know how to do that) and when clicking them I did see the tip pop up, which is good... but leaves me wondering if I would have discovered that while reading it.  Most likely, my lazy overworked tutorial-reading self would have copied and pasted commands into a terminal without noticing the `+` icon.

**Comment by bjacob** [comment]
I also wonder if the "Touring the Tracy Profiler UI" section is worth keeping. It feels you went out of your way to preserve that content that I had added years ago (the screenshots say 2020!) and leaving aside the question of how the UI changed since then, I just wonder if their benefits outweigh the cost of making this document bulkier.  Either way we need to be consistent in our bulk / explicitness tradeoffs:  above we were making the rather extreme trade-off of hiding crucial information ("need to run as root") behind `+` icons.

File: docs/website/docs/developers/performance/profiling-with-tracy.md (hunk 1 above)
Comment (ScottTodd) [reply docs/website/docs/developers/performance/profiling-with-tracy.md:43]: > but leaves me wondering if I would have discovered that while reading it

I was hoping the highlights would help draw attention there, but yeah... I share this concern. Thinking I might restructure a bit more. Could introduce some repetition in content (dedicated section _and_ `+` icon annotations)

Comment (ScottTodd) [reply docs/website/docs/developers/performance/profiling-with-tracy.md:43]: Added more information upfront about what sampling requires.

**Comment by ScottTodd** [comment]
> I also wonder if the "Touring the Tracy Profiler UI" section is worth keeping. It feels you went out of your way to preserve that content that I had added years ago (the screenshots say 2020!) and leaving aside the question of how the UI changed since then, I just wonder if their benefits outweigh the cost of making this document bulkier. Either way we need to be consistent in our bulk / explicitness tradeoffs: above we were making the rather extreme trade-off of hiding crucial information ("need to run as root") behind `+` icons.

I'm fine with keeping that section. I think it's a useful substitute for a live demo since it provides a step by step walkthrough of how to use the program. I actually want to expand on that section to talk about comparing trace files and looking through statistics, but left that for future work.

File: docs/website/docs/developers/performance/profiling-with-tracy.md (hunk 1 above)
Comment (ScottTodd) [reply docs/website/docs/developers/performance/profiling-with-tracy.md:43]: Removed the `+` icons in favor of code block comments. Should be easier to copy/paste now: 
image

**Review by bjacob — APPROVED** [review]

**Pull request merged**